Add more context and text around ACK frequency (#3706)
quicwg/base-drafts: draft-ietf-quic-transport.md

value in determining timeouts for timer-based retransmission, as detailed in
Section 5.2.1 of {{QUIC-RECOVERY}}.

Since packets containing only ACK frames are not congestion controlled, an
endpoint MUST NOT send more than one such packet in response to receiving an
ack-eliciting packet.

An endpoint MUST NOT send a non-ack-eliciting packet in response to a
non-ack-eliciting packet, even if there are packet gaps which precede the
received packet. This avoids an infinite feedback loop of acknowledgements,

> which could prevent the connection from ever becoming idle. Note that
> non-ack-eliciting packets are eventually acknowledged when the endpoint sends an
> ACK frame in response to other events.

Reviewer: ```suggestion
which could prevent the connection from ever becoming idle.
Non-ack-eliciting packets are eventually acknowledged when the endpoint sends an
ACK frame in response to other events.
```

the IP header SHOULD be acknowledged immediately, to reduce the peer's response
time to congestion events.

> The algorithms in {{QUIC-RECOVERY}} are resilient to receivers that do not

Reviewer: do we want to add "generally" to "resilient"?  Or, "in principle"?  Or "in theory"?

ack-eliciting packet has been received to bundle an ACK frame with outgoing
frames.

### Acknowledgement Frequency

A receiver determines how frequently to send acknowledgements in response to

> ack-eliciting packets. This determination involves a tradeoff however.

Reviewer: ```suggestion
ack-eliciting packets. This determination involves a tradeoff.
```

> Endpoints rely on timely acknowledgment to detect loss; see Section X of

Reviewer: We'll need a value for X.

```suggestion
Endpoints rely on timely acknowledgment to detect loss; see Section 5 of
```

Author: Gah! Sorry, meant to replace that.

{{QUIC-RECOVERY}}. Window-based congestion controllers, such as the one in

> Section X of {{QUIC-RECOVERY}}, rely on acknowledgments to manage their sending

Reviewer: ```suggestion
Section 6 of {{QUIC-RECOVERY}}, rely on acknowledgments to manage their sending
```

> Section X of {{QUIC-RECOVERY}}, rely on acknowledgments to manage their sending
> rate. In both cases, delaying acknowledgment can adversely affect performance.

Reviewer 2: ```suggestion
Section X of {{QUIC-RECOVERY}}, rely on acknowledgments to open up congestion window,
allowing more packets to be sent, as well as increase the congestion window. In both
cases, delaying acknowledgments can adversely affect performance.
```

> On the other hand, reducing the frequency of acknowledgement packets reduces

Reviewer 2: Do you mean ACK-only packets?  Otherwise this could be read to include ACK frames bundled with data, which I don't believe is the intent.

packet processing cost at both endpoints. It can also improve connection
throughput on severely asymmetric links; see Section 3 of {{?RFC3449}}.

A receiver SHOULD send an ACK frame after receiving at least two ack-eliciting

> packets. Better knowledge of network conditions or further research and
> experimentation might suggest alternative acknowledgment strategies with better
> performance characteristics. This recommendation is general in nature and

Reviewer 2: ```suggestion
packets. Knowledge of how frequently the peer's congestion controller needs feedback,
network conditions or further research and experimentation might suggest alternative
acknowledgment strategies with better performance characteristics. This
recommendation is general in nature and
```

> consistent with recommendations for TCP endpoint behavior {{?RFC5681}}.

Reviewer: I think maybe we can swap the last two sentences, so that "this recommendation" is better bound to the first sentence.

> A receiver MAY process multiple packets before sending any ACK frames in

Reviewer 2: ```suggestion
A receiver MAY process multiple available packets before determining whether to sending an ACK frame in
```

> response. This allows for a receiver to process packets already queued for
> processing before determining whether to send an acknowledgement.

Reviewer 2: ```suggestion
response.
```

ack-eliciting packet has been received to bundle an ACK frame with outgoing
frames.

### Acknowledgement Frequency

A receiver determines how frequently to send acknowledgements in response to
ack-eliciting packets. This determination involves a tradeoff.

Endpoints rely on timely acknowledgment to detect loss; see Section 5 of
{{QUIC-RECOVERY}}. Window-based congestion controllers, such as the one in

> Section 6 of {{QUIC-RECOVERY}}, rely on acknowledgments to manage their sending

Reviewer 2: I would still like to be clearer here and I don't really like "sending rate" in the context of window-based controllers.  I made some specific suggestions before.  Can you take another look at them?

Author: It's still quite accurate, IMO. Ultimately, increasing the congestion window increases the sending rate. That is the higher order point here -- that reducing ack frequency reduces the throughput of the sender.

Author: I've changed it to "manage their congestion window"

Reviewer 2: Thanks, I think that's better, but I also think it doesn't highlight a key point, which is if a receiver delays sending an ACK for Xms, the sender may be blocked from sending for Xms.

Except for slow start, I don't think the congestion window itself is the issue, but rather the available congestion window.

> rate. In both cases, delaying acknowledgment can adversely affect performance.

Reviewer 2: ```suggestion
rate. In both cases, delaying acknowledgments can adversely affect performance.
```

On the other hand, reducing the frequency of packets that carrying only

> acknowledgements reduces packet processing cost at both endpoints. It can also

Reviewer 2: Q: Doesn't it reduce packet serialization and sending costs at the receiver and packet processing costs at the sender?  In practice, the cost of sending ACK-only packets is commonly worse, so it seems worth mentioning.

Author: That is a detail, which will need further substantiation. The point here is to note _some_ benefits, to point out that there are some. The specifics will depend on your particular platform.

Reviewer 2: Now that I read this again, I'm actually not sure what you mean.  Do you mean packet processing cost as in the cost of receiving a large number of data packets, so fewer ACK-only packets reduces that cost?

I was thinking you were saying it reduces the number of ACKs that have to be received and processed.

Either way, I have a small suggestion.
```suggestion
acknowledgements reduces packet transmission and processing cost at both endpoints. It can also
```

ack-eliciting packet has been received to bundle an ACK frame with outgoing
frames.

### Acknowledgement Frequency

A receiver determines how frequently to send acknowledgements in response to
ack-eliciting packets. This determination involves a tradeoff.

Endpoints rely on timely acknowledgment to detect loss; see Section 5 of
{{QUIC-RECOVERY}}. Window-based congestion controllers, such as the one in
Section 6 of {{QUIC-RECOVERY}}, rely on acknowledgments to manage their
congestion window. In both cases, delaying acknowledgments can adversely affect
performance.

> On the other hand, reducing the frequency of packets that carrying only

Reviewer 2: ```suggestion
On the other hand, reducing the frequency of packets that carry only
```

ack-eliciting packet has been received to bundle an ACK frame with outgoing
frames.

### Acknowledgement Frequency

A receiver determines how frequently to send acknowledgements in response to
ack-eliciting packets. Under normal circumstances, reducing the frequency of
acknowledgement packets can improve connection and endpoint performance in
several ways, such as reducing computational cost at both endpoints, and

> improving connection throughput on severely asymmetric links.

Reviewer: You need to phrase this more as a trade-off.  This is concentrated on the benefits of reducing acknowledgments.  But the introduction here also needs to recognize that generating acknowledgments in a timely fashion is critical for performance.

> However, there are undesirable consequences to a receiver simply reducing the

Reviewer: If you balance the first paragraph out, then you don't need a "however" here.  You can just talk in more detail about the advantages of more frequent acknowledgment.

acknowledgement frequency, especially to an arbitrary fixed value. A sender
relies on receipt of acknowledgements to determine the amount of data in flight
and to detect loss; see {{QUIC-RECOVERY}}. Loss detection can be delayed with
late acknowledgements. Window-based congestion controllers, such as the one
defined in {{QUIC-RECOVERY}}, rely on receipt of acknowledgments to increase

> their sending rate. Consequently, a connection can suffer performance penalties

Reviewer: If you move the general statement at the end of this to the first paragraph, then you can probably shorten this paragraph to: 

> Endpoints rely on timely acknowledgment to detect loss; see Section X of {{QUIC-RECOVERY}}.  Window-based congestion controllers, such as the one in Section X of {{QUIC-RECOVERY}} rely on acknowledgments to manage their sending rate.  In both cases, delaying acknowledgment can adversely affect performance.

We don't need the specific dig regarding the "arbitrary fixed value".

when acknowledgements are not sent frequently enough.

Further research and experimentation will yield effective mechanisms to balance
this tradeoff. Meanwhile, this document provides the following guidance:
{{sending-acknowledgements}}, a receiver SHOULD generate an ACK frame for at

> least every second ack-eliciting packet. This recommendation is in keeping with

Reviewer: A receiver SHOULD send an ACK frame after receiving at least two ack-eliciting packets.  Better knowledge of network conditions or further research and experimentation might suggest alternative acknowledgment strategies with better performance characteristics.  This recommendation is general in nature and consistent with recommendations for TCP endpoint behavior {{?RFC5681}}.

data sent by the server protected by the 1-RTT keys.

### PADDING Frames Consume Congestion Window

Packets containing PADDING frames are considered to be in flight for congestion
control purposes {{QUIC-RECOVERY}}. Sending only PADDING frames might cause the
sender to become limited by the congestion controller with no acknowledgments
forthcoming from the receiver. Therefore, a sender SHOULD ensure that other

> frames are sent periodically in addition to PADDING frames to elicit acknowledgments from

Reviewer: Check line length.

data sent by the server protected by the 1-RTT keys.

### PADDING Frames Consume Congestion Window

Packets containing PADDING frames are considered to be in flight for congestion

> control purposes {{QUIC-RECOVERY}}. Sending only PADDING frames might cause the
> sender to become limited by the congestion controller with no acknowledgments
> forthcoming from the receiver. Therefore, a sender SHOULD ensure that other
> frames are sent periodically in addition to PADDING frames to elicit
> acknowledgments from the receiver.

Reviewer: ```suggestion
control purposes {{QUIC-RECOVERY}}. Packets containing only PADDING frames
therefore consume congestion window but do not necessarily generate
acknowledgments that will open the congestion window. Therefore, a sender SHOULD
ensure that other frames are sent periodically in addition to PADDING frames to
elicit acknowledgments from the receiver.
```

A maybe-improvement.

Reviewer 2: I'd remove "necessarily" from your suggestion, but otherwise LGTM.

Reviewer: Suggestion updated.

data sent by the server protected by the 1-RTT keys.

> ### Other Considerations

Reviewer: I'm sure that we can do better than "Other Considerations" here.

```suggestion
### PADDING Frames Consume Congestion Window
```

Packets containing PADDING frames are considered to be in flight for congestion
control purposes {{QUIC-RECOVERY}}. Sending only PADDING frames might cause the
sender to become limited by the congestion controller with no acknowledgments
forthcoming from the receiver. Therefore, a sender SHOULD ensure that other

> frames are sent in addition to PADDING frames to elicit acknowledgments from

Reviewer: ```suggestion
frames are sent periodically in addition to PADDING frames to elicit acknowledgments from
```
